### PR TITLE
Temporary global variable aliasing within kernels #574

### PR DESCRIPTION
**Pull request by archibate**
Related issue id = #574

### CODE REVIEW

**Comment by archibate** [comment]
Why so mess? Sorry, I'll fix this.

**Comment by yuanming-hu** [comment]
I cleared the extra commits to make the changeset visible. I suggest sticking to `ti.static` instead of introducing a new keyword `ti.just`.

**Comment by archibate** [comment]
spdlog/fmt/bundled/color.h file not found again.

**Comment by yuanming-hu** [comment]
You probably need to update the spdlog submodule.

**Comment by archibate** [comment]
Tried `git submodule update`, didn't work. Can we have an official clone of spdlog? Maybe color.h not found is because they moved it somewhere else?

**Comment by yuanming-hu** [comment]
For some reason, the `spdlog` submodule version was modified. Now it should work. Please try again.

**Comment by archibate** [comment]
Works! Thank you! So we should make a clone of spdlog for fear of their rapid changes?

**Comment by yuanming-hu** [comment]
I think we have already specified commit cf6f1dd01e660d5865d68bf5fa78f6376b89470a of spdlog.

File: python/taichi/lang/impl.py [hunk 1]
```diff
@@ -286,13 +286,18 @@ def indices(*x):
 index = indices
 
 
-def static(x):
+def static(*xs):
+  if len(xs) > 1: # for python-ish pointer assign: x, y = ti.static(y, x)
+    return tuple(static(x) for x in xs)
+  else:
+    x = xs[0] # len != 0 -> Error
```
Comment (yuanming-hu) [inline python/taichi/lang/impl.py]: let throw `ValueError` here if `len(xs)` == 0

File: python/taichi/lang/impl.py [hunk 2]
```diff
@@ -286,13 +286,19 @@ def indices(*x):
...
+  else:
+    x = xs[0]
   import taichi as ti
-  assert get_runtime(
-  ).inside_kernel, 'ti.static can only be used inside Taichi kernels'
+  assert get_runtime().inside_kernel, 'ti.static can only be used inside Taichi kernels'
   assert isinstance(
       x, (bool, int, float, range, list, tuple, ti.ndrange, ti.GroupedNDRange)
-  ), 'Input to ti.static must have compile-time constant values, instead of {}'.format(
+  ) or (isinstance(x, ti.lang.expr.Expr) and x.ptr.is_global_var()
```
Comment (yuanming-hu) [inline python/taichi/lang/impl.py]: Let's support `ti.Matrix` here as well :-)

**Comment by yuanming-hu** [comment]
@KLozes @archibate thanks for proposing and implementing this.

If we can add **a few test cases** and **a note on how to use this in the documentation (Taichi > Advanced programming > Syntax sugars > Static assignment.rst)**, that would greatly benefit future developers and users.

**Comment by KLozes** [comment]
Yes, thanks @archibate ! Wish I had more time on my hands to make contributions myself.  Also, just added some docs for this feature.  I'll make some tests now too

File: python/taichi/lang/impl.py [hunk 3]
```diff
@@ -286,15 +286,21 @@ def indices(*x):
 index = indices
 
 
-def static(x):
+def static(x, *xs):
+  if len(xs): # for python-ish pointer assign: x, y = ti.static(y, x)
+    return [static(x)] + [static(x) for x in xs]
```
Comment (archibate) [inline python/taichi/lang/impl.py]: Sorry for breaking your ccbb81a :) I thought I've found a more elegant way to deal with it like other function does.
I mean, to use python built-in argument number dismatch error like `ti.grouped` will do.
If you don't like this, pls feel free to revert it as you want.

File: tests/python/test_ptr_assign.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    assert a[2, 0] == x + y
+
+@ti.all_archs
+def test_pythonish_swap():
+  a = ti.var(dt=ti.f32, shape=())
+  b = ti.var(dt=ti.f32, shape=())
+
+  @ti.kernel
+  def func(x: ti.f32, y: ti.f32):
+    u, v = ti.static(b, a)
```
Comment (archibate) [inline tests/python/test_ptr_assign.py]: Got `UnboundLocalError: local variable 'b' referenced before assignment` when trying
```py
a, b = ti.static(b, a)
```
here, any idea? Have to sleep now, good night~

File: python/taichi/lang/impl.py [hunk 5]
```diff
@@ -286,15 +286,21 @@ def indices(*x):
...
-  assert isinstance(
-      x, (bool, int, float, range, list, tuple, ti.ndrange, ti.GroupedNDRange)
-  ), 'Input to ti.static must have compile-time constant values, instead of {}'.format(
-      type(x))
-  return x
+  if isinstance(x, (bool, int, float, range, list, tuple, ti.ndrange, ti.GroupedNDRange)):
+    return x
+  elif isinstance(x, ti.lang.expr.Expr) and x.ptr.is_global_var():
+    return x
+  elif isinstance(x, ti.Matrix) and x.entries[0].ptr.is_global_var():
```
Comment (yuanming-hu) [inline python/taichi/lang/impl.py]: Let's use `Matrix.is_global`

File: tests/python/test_ptr_assign.py (hunk 4 above)
Comment (yuanming-hu) [reply tests/python/test_ptr_assign.py]: No worries about this case. Assigning to global variables is forbidden anyway.

File: python/taichi/lang/impl.py (hunk 3 above)
Comment (yuanming-hu) [reply python/taichi/lang/impl.py]: Not sure if I get your point correctly

Comment (archibate) [reply python/taichi/lang/impl.py]: I mean, `ti.static(x, *xs)` have the same error message with the `ti.static(x)` instead a custom ValueError.

Comment (yuanming-hu) [reply python/taichi/lang/impl.py]: I see. But keep in mind that having recursion here will lead to confusing stack backtrace if the final entry of `ti.static` leads to any error...

Comment (archibate) [reply python/taichi/lang/impl.py]: Also pls think if we want `ti.static(range(2), range(3))` as syntax correct.

Comment (yuanming-hu) [reply python/taichi/lang/impl.py]: This is really a corner case - I would say the users should avoid confusing expressions like these. Let's allow this for now - the users should use it at their own risk.

**Pull request merged**